### PR TITLE
Integration tests for short-circuit response

### PR DESCRIPTION
**Pull request by k-wall**
### Type of change

- Enhancement / new feature

### Description

Adding tests cases that cover synchronous/asynchronous short circuit responses, with and without connection close.

Why: There's some tricky coding around the handling of deferred responses.  Let's improve test coverage to build confidence that everything is working and stays that way.

### Additional Context

To test this, I need to the ability for the mock client to sense whether its channel had been closed on not. This necessitated a refactoring of the mock client so that it allows longer conversations so that observations of the channel's connected state could be made, or even await a state change.

I took a different approach to that of https://github.com/kroxylicious/kroxylicious/pull/507 in that I have a `multiRequestClient` which allows > 1 calls to requestClient.getSync.   I *think* this might be more useful, especially when testing filters that have state.

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [x] Write tests
- [x] Make sure all tests pass
- [ ] Review performance test results. Ensure that any degradations to performance numbers are understood and justified.
- [ ] Make sure all Sonar-Lint warnings are addressed or are justifiably ignored.
- [ ] Update documentation
- [ ] Reference relevant issue(s) and close them after merging
- [ ] For user facing changes, update CHANGELOG.md (remember to include changes affecting the API of the test artefacts too).

### CODE REVIEW

File: kroxylicious-test-tools/src/main/java/io/kroxylicious/test/client/KafkaClientHandler.java
Comment (SamBarker) [inline kroxylicious-test-tools/src/main/java/io/kroxylicious/test/client/KafkaClientHandler.java:50]: Do the exceptions still bubble up somewhere or is this silently ignoring them?

File: kroxylicious-test-tools/src/main/java/io/kroxylicious/test/client/KafkaClientHandler.java
Comment (SamBarker) [inline kroxylicious-test-tools/src/main/java/io/kroxylicious/test/client/KafkaClientHandler.java:84]: I don't have a good picture of when the channel would be active or inactive but it feels strange to gate entering the loop on if the channel is active or not. I would have thought we would do that inside the loop.
```suggestion
            while (queue.peek() != null) {
                if (ctx.channel().isActive()) {
                    DecodedRequestFrame<?> msg = queue.removeFirst();
                    ctx.writeAndFlush(msg).addListener(c -> {
                        if (msg.body() instanceof ProduceRequest && ((ProduceRequest) msg.body()).acks() == 0) {
                            msg.getResponseFuture().complete(null);
                        }
                    });
                }
            }
        });
```

File: kroxylicious-test-tools/src/test/java/io/kroxylicious/test/tester/KroxyliciousTestersTest.java
Comment (SamBarker) [inline kroxylicious-test-tools/src/test/java/io/kroxylicious/test/tester/KroxyliciousTestersTest.java:138]: Should we have a version of the test that verifies the singleRequest client is closed after its request is complete.

File: integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java
Comment (SamBarker) [inline integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java:219]:
```suggestion
                    .untilAsserted(() -> assertThat(requestClient.isOpen()).isNotEqualTo(closeConnection));
```

I'm tempted to say we should invert the `closeConnection` variable so this becomes 
```suggestion
                    .untilAsserted(() -> assertThat(requestClient.isOpen()).isEqualTo(expectConnectionOpen));
```

**Comment by robobario** [comment]
Is there any value in keeping the one-shot mode? Maybe instead of having these modes it should only offer the long-lived channel, maybe with a `closeConnection()` type method on the client if a test wanted to re-connect.

Also we should decide on how to approach https://github.com/kroxylicious/kroxylicious/issues/526, I think the KroxyliciousTester should be responsible for redundantly closing any clients it has created, but this has the complication outlined in the issue.

File: integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java
Comment (SamBarker) [inline integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java:216]: I think [`asInstanceOf`](https://assertj.github.io/doc/#assertj-core-3.13.0-asInstanceOf) could help here.

**Review by SamBarker — APPROVED** [review]
Would have been nice to separate the test changes from the client refactoring. 

I'm with @robobario I'm not sure about the merits of a single request vs multi request client but this feels like progress and satisfies a use case. 

For me the singleRequest client shouldn't need to be closed it should just disconnect and clean up but thats not really in scope here. I don't think.

File: kroxylicious-test-tools/src/main/java/io/kroxylicious/test/client/KafkaClientHandler.java
Comment (robobario) [reply kroxylicious-test-tools/src/main/java/io/kroxylicious/test/client/KafkaClientHandler.java:84]: ^ would be an infinite loop of peeking and checking active if the channel died. But yeah maybe it should drain the queue and if the channel is closed complete the future exceptionally?

File: kroxylicious-test-tools/src/main/java/io/kroxylicious/test/client/KafkaClientHandler.java
Comment (SamBarker) [inline kroxylicious-test-tools/src/main/java/io/kroxylicious/test/client/KafkaClientHandler.java:80]: Can we extract this special case somehow? Maybe it should be:
```suggestion
                        if (!msg.ackRequired) {
                            msg.getResponseFuture().complete(null);
                        }
```

Doesn't need to happen as part of this PR.

File: kroxylicious-test-tools/src/main/java/io/kroxylicious/test/client/KafkaClientHandler.java
Comment (SamBarker) [reply kroxylicious-test-tools/src/main/java/io/kroxylicious/test/client/KafkaClientHandler.java:84]: Hmm I certainly think we should be explicit about what happens if the connection is closed. 

if the connection is closed during the draining of the queue we would end up with logs of exceptions I think so maybe need to test the channel is active and break if not during the loop as well.

File: integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java
Comment (k-wall) [reply integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java]: Renamed to `RejectingCreateTopic`.  We already had a Prefixing filter, so I went that way.

File: kroxylicious-test-tools/src/main/java/io/kroxylicious/test/client/KafkaClientHandler.java
Comment (k-wall) [reply kroxylicious-test-tools/src/main/java/io/kroxylicious/test/client/KafkaClientHandler.java:50]: This is Netty's inbound exception handling.  My motivation for removing the printStackTrace was incorrect.  I wrongly believed I would see exception related to the close of the channel from the server side.  That's not correct, the channel's closure is graceful and there's no exception.

I'll change the printStackTrace into a warn.  We don't expect to follow this path.

File: kroxylicious-test-tools/src/main/java/io/kroxylicious/test/client/KafkaClientHandler.java
Comment (k-wall) [reply kroxylicious-test-tools/src/main/java/io/kroxylicious/test/client/KafkaClientHandler.java]: Done.

File: kroxylicious-test-tools/src/main/java/io/kroxylicious/test/client/KafkaClientHandler.java
Comment (k-wall) [reply kroxylicious-test-tools/src/main/java/io/kroxylicious/test/client/KafkaClientHandler.java]: I've refactored to propagate the failed write exception to the response future.

File: integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java
Comment (k-wall) [reply integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java]: I like the fact that the test is in terms of the public API.  I mean the FilterResultBuilder talks in terms of the `withCloseConnection`, so the test should too. That made me prefer to keep the test's interface the same.  

I extracted a variable `expectConnectionOpen` to reduce the mental gymnastics.

File: kroxylicious-test-tools/src/test/java/io/kroxylicious/test/tester/KroxyliciousTestersTest.java
Comment (k-wall) [reply kroxylicious-test-tools/src/test/java/io/kroxylicious/test/tester/KroxyliciousTestersTest.java]: I got rid of the single request client.

**Comment by k-wall** [comment]
> Would have been nice to separate the test changes from the client refactoring.

I'm sorry if that's made the review harder.  I grew the changes to the test framework with the test itself.  This was partly expediency and partly just unfamiliarity with the implementation of the mock framework.

This PR didn't quite cross the line where I split a PR into two and park the first behind the second.

> 
> I'm with @robobario I'm not sure about the merits of a single request vs multi request client but this feels like progress and satisfies a use case.

I ditched the single request case.

> 
> For me the singleRequest client shouldn't need to be closed it should just disconnect and clean up but thats not really in scope here. I don't think.

File: kroxylicious-test-tools/src/main/java/io/kroxylicious/test/tester/DefaultKroxyliciousTester.java
Comment (robobario) [inline kroxylicious-test-tools/src/main/java/io/kroxylicious/test/tester/DefaultKroxyliciousTester.java:113]: I wonder if the naming is misleading (the MockServer is a mock broker, but the client is doing real client work). Maybe `lowLevelClient`, `simpleClient`, `testClient`?

**Review by robobario — APPROVED** [review]
LGTM

File: kroxylicious-test-tools/src/main/java/io/kroxylicious/test/tester/DefaultKroxyliciousTester.java
Comment (SamBarker) [reply kroxylicious-test-tools/src/main/java/io/kroxylicious/test/tester/DefaultKroxyliciousTester.java:113]: yeah @robobario is right. I don't think calling it a mock client is helpful

File: CHANGELOG.md
Comment (SamBarker) [inline CHANGELOG.md:9]: again this makes it sound like a client that mocks things out not a stubbed broker.

**Review by SamBarker — APPROVED** [review]
I'd like `mockClient` renamed to something else (its also not a Mock but a Stub broker) but I don't feel the need of another review cycle.

File: kroxylicious-test-tools/src/main/java/io/kroxylicious/test/tester/DefaultKroxyliciousTester.java
Comment (k-wall) [reply kroxylicious-test-tools/src/main/java/io/kroxylicious/test/tester/DefaultKroxyliciousTester.java:113]: Yes, you are both right.  I'll go with simpleTestClient.

**Pull request merged**
